### PR TITLE
feature(wyh): add guided cost algorithm

### PR DESCRIPTION
**Pull request by Weiyuhong-1998**
## Description

guided cost Algorithm．support cartpole, lunarlander, hopper, mujoco, halfcheetah
Doc link: https://github.com/opendilab/DI-engine-docs/pull/44

## Check List

- [ ] merge the latest version source branch/repo, and resolve all the conflicts
- [ ] pass style check
- [ ] pass all the tests

### CODE REVIEW

File: ding/entry/serial_entry_max_entropy.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+        # Collect data by default config n_sample/n_episode
+        new_data = collector.collect(train_iter=learner.train_iter, policy_kwargs=collect_kwargs)
+        train_data = copy.deepcopy(new_data)
+        expert_data = expert_collector.collect(train_iter=learner.train_iter, policy_kwargs=collect_kwargs)
+        for i in range(len(new_data)):
+            device_1 = new_data[i]['obs'].device
+            device_2 = expert_data[i]['obs'].device
+            probs = F.softmax(new_data[i]['logit'], dim = -1)
+            #new_data[i]['prob'] = torch.tensor(probs[new_data[i]['action']]).to(device_1)
+            new_data[i]['prob'] = probs[new_data[i]['action']].to(device_1)
```
Comment (PaParaZz1) [inline ding/entry/serial_entry_max_entropy.py:134]: prob calculation can be placed in `estimate` method of `BaseRewardModel`

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: cartpole_ppo_offpolicy_config.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    ),
+    env_manager=dict(type='base'),
+    policy=dict(type='ppo_offpolicy'),
+    reward_model=dict(type='max_entropy'),
+)
+cartpole_ppo_offpolicy_create_config = EasyDict(cartpole_ppo_offpolicy_create_config)
+create_config = cartpole_ppo_offpolicy_create_config
+
+if __name__ == "__main__":
+    serial_pipeline_max_entropy([main_config, create_config], seed=0)
```
Comment (Will-Nie) [inline cartpole_ppo_offpolicy_config.py:62]: Why do not you create a trex config? Simply copy this config while adding the reward_model part

File: cartpole_ppo_offpolicy_config.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+cartpole_ppo_offpolicy_config = EasyDict(cartpole_ppo_offpolicy_config)
+main_config = cartpole_ppo_offpolicy_config
+cartpole_ppo_offpolicy_create_config = dict(
+    env=dict(
+        type='cartpole',
+        import_names=['dizoo.classic_control.cartpole.envs.cartpole_env'],
+    ),
+    env_manager=dict(type='base'),
+    policy=dict(type='ppo_offpolicy'),
+    reward_model=dict(type='max_entropy'),
```
Comment (Will-Nie) [inline cartpole_ppo_offpolicy_config.py:56]: Why do you put the 'type' of this reward model in create config section? what about moving it to the main config to unify styles

File: ding/entry/serial_entry_max_entropy.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+        env=expert_collector_env,
+        policy=expert_policy.collect_mode,
+        tb_logger=tb_logger,
+        exp_name=cfg.exp_name
+    )
+    evaluator = BaseSerialEvaluator(
+        cfg.policy.eval.evaluator, evaluator_env, policy.eval_mode, tb_logger, exp_name=cfg.exp_name
+    )
+    replay_buffer = create_buffer(cfg.policy.other.replay_buffer, tb_logger=tb_logger, exp_name=cfg.exp_name)
+    expert_buffer = create_buffer(cfg.policy.other.replay_buffer, tb_logger=tb_logger, exp_name=cfg.exp_name)
```
Comment (Will-Nie) [inline ding/entry/serial_entry_max_entropy.py:97]: By modifying  the load_expert_data function from reward model, you can either collect expert data from a policy (see collect_demo_data from ding.entry.application_entry.py) or load an offline dataset. This can help you further polish the entry and allow more flexibility to collect data.

File: ding/entry/serial_entry_max_entropy.py [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+            if stop:
+                break
+        # Collect data by default config n_sample/n_episode
+        new_data = collector.collect(train_iter=learner.train_iter, policy_kwargs=collect_kwargs)
+        train_data = copy.deepcopy(new_data)
+        expert_data = expert_collector.collect(train_iter=learner.train_iter, policy_kwargs=collect_kwargs)
+        replay_buffer.push(new_data, cur_collector_envstep=collector.envstep)
+        expert_buffer.push(expert_data, cur_collector_envstep=expert_collector.envstep)
+        # Learn policy from collected data
+        for i in range(cfg.reward_model.update_per_collect):
```
Comment (Will-Nie) [inline ding/entry/serial_entry_max_entropy.py:132]: Why not set this update_per_collect to be an independent hyper-parameter?

File: ding/entry/serial_entry_max_entropy.py [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+                    "You can modify data collect config, e.g. increasing n_sample, n_episode."
+                )
+                break
+            learner.train(train_data, collector.envstep)
+            if learner.policy.get_attribute('priority'):
+                replay_buffer.update(learner.priority_info)
+        if cfg.policy.on_policy:
+            # On-policy algorithm must clear the replay buffer.
+            replay_buffer.clear()
+        dirname = cfg.exp_name+'/reward_model'
```
Comment (Will-Nie) [inline ding/entry/serial_entry_max_entropy.py:154]: Please add this path to config

File: cartpole_ppo_offpolicy_config.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
+from easydict import EasyDict
+from ding.entry import serial_pipeline_max_entropy
+
+cartpole_ppo_offpolicy_config = dict(
+    exp_name='cartpole_ppo_offpolicy2',
```
[the remark this answers is not in the document]
Comment (Will-Nie) [reply cartpole_ppo_offpolicy_config.py:5]: Can you please rename all the configs? and after finishing this, please press 'resolve'

File: ding/reward_model/max_entropy_irl_model.py [hunk 8]
```diff
@@ -70,24 +83,25 @@ def train(self, expert_demo: torch.Tensor, samp: torch.Tensor, iter, step):
         prob = samp['prob'].unsqueeze(-1)
         #print(len(cost_samp))
```
Comment (Will-Nie) [inline ding/reward_model/max_entropy_irl_model.py:84]: I suppose it will be nice to print some logs just as RL training part. Maybe it is nice to uncomment some important outputs to let people see their IRL part training. After this, please delete all irrelevant comments.

File: ding/entry/serial_entry_max_entropy.py [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+        input_cfg: Union[str, Tuple[dict, dict]],
+        seed: int = 0,
+        env_setting: Optional[List[Any]] = None,
+        model: Optional[torch.nn.Module] = None,
+        expert_model: Optional[torch.nn.Module] = None,
+        max_iterations: Optional[int] = int(1e10),
+) -> 'Policy':  # noqa
+    """
+    Overview:
+        Serial pipeline sqil entry: we create this serial pipeline in order to\
```
Comment (Will-Nie) [inline ding/entry/serial_entry_max_entropy.py:32]: polish comments

File: ding/entry/serial_entry_max_entropy.py [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+from ding.worker import BaseLearner, SampleCollector, BaseSerialEvaluator, BaseSerialCommander, create_buffer, \
+    create_serial_collector
+from ding.config import read_config, compile_config
+from ding.policy import create_policy, PolicyFactory
+from ding.utils import set_pkg_seed, save_file
+from ding.reward_model import create_reward_model
+import copy
+import os
+
+def serial_pipeline_max_entropy(
```
Comment (Will-Nie) [inline ding/entry/serial_entry_max_entropy.py:22]: maybe serial_pipeline_gcl

File: ding/policy/sac.py [hunk 11]
```diff
@@ -162,6 +162,7 @@ class SACPolicy(Policy):
             # You can use either "n_sample" or "n_episode" in actor.collect.
             # Get "n_sample" samples per collect.
             # Default n_sample to 1.
+            use_logit=False,
```
Comment (PaParaZz1) [inline ding/policy/sac.py:165]: add comment for this argument，and our variable name should be like `use_xxx`, `is_xxx`，`collector_logit` can be better

File: ding/reward_model/guided_cost_reward_model.py [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+            nn.ReLU(),
+            nn.Linear(hidden_size, output_size),
+        )
+
+    def forward(self, x):
+        return self.net(x)
+
+
+@REWARD_MODEL_REGISTRY.register('guided_cost')
+class GuidedCostModel(BaseRewardModel):
```
Comment (PaParaZz1) [inline ding/reward_model/guided_cost_reward_model.py:39]: `GuidedCostRewardModel`

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: dizoo/mujoco/config/walker2d_gcl_config.py [hunk 13]
```diff
@@ -0,0 +1,72 @@
+from copy import deepcopy
+from ding.entry import serial_pipeline, serial_pipeline_guided_cost
```
Comment (Will-Nie) [inline dizoo/mujoco/config/walker2d_gcl_config.py:2]: polish config, delete unnecessary imports. eg, deepcopy, serial_pipeline

File: ding/entry/serial_entry_guided_cost.py [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+import torch
+import logging
+from functools import partial
+from tensorboardX import SummaryWriter
+import numpy as np
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from ding.envs import get_vec_env_setting, create_env_manager
+from ding.worker import BaseLearner, SampleCollector, BaseSerialEvaluator, BaseSerialCommander, create_buffer, \
```
Comment (Will-Nie) [inline ding/entry/serial_entry_guided_cost.py:13]: BaseSerialEvaluator has been changed into InteractionSerialEvaluator

**Review by PaParaZz1 — APPROVED** [review]
